### PR TITLE
feat: add validator run event to allow plugins to perform custom validation

### LINKED ISSUE #2183: Custom validation
## Search Terms

validate
validation

## Problem

I want to create a custom validation check but that doesn't currently seem possible.

## Suggested Solution

Allow plugins access to an "on validate" event. They can then run their own logic and return either `ture` or `false`.

### PR DESCRIPTION
**Pull request by RebeccaStevens**
fix #2183

### CODE REVIEW

**Comment by Gerrit0** [comment]
I don't think I want the validator to be a separate thing - it's not nearly complicated enough to deserve that... I'd rather just add the validate event to `ApplicationEvents` and emit it on the application, then if any validator logs or listener to that method logs an error, TypeDoc will exit with the ValidationError exit code.

If it does deserve pulling out into its own class someday - I'd still want to avoid the component and emit that event from the application I think... I'm not convinced that there's value for plugins in seeing a new class.

**Comment by RebeccaStevens** [comment]
@Gerrit0 I've reworked this PR to address those issue :)

**Comment by RebeccaStevens** [comment]
When working on my plugin that relies on this PR, `reflection.comment.getTags("@param")` always return an empty array. I'm not having this issue with any other tags. Am I missing something for this tag or is this a bug?

File: src/lib/utils/events.ts [hunk 1]
```diff
@@ -433,6 +436,12 @@ export class EventDispatcher {
         context?: any,
         priority?: number
     ): this;
+    on(
```
Comment (Gerrit0) [inline src/lib/utils/events.ts:439]: I'd rather not do this for now - better to have none of the events typed than some of them. Eventually it'd be nice to have autocomplete, but that'll probably not come until removing `@Component`...

File: src/lib/application.ts [hunk 2]
```diff
@@ -107,6 +107,12 @@ export class Application extends ChildableComponent<
      */
     static readonly EVENT_BOOTSTRAP_END = ApplicationEvents.BOOTSTRAP_END;
 
+    /**
+     * Emitted when validation is being run.
+     * The listener will be given an instance of {@link ProjectReflection}.
+     */
+    static readonly EVENT_VALIDATION_RUN = ApplicationEvents.VALIDATION_RUN;
```
Comment (Gerrit0) [inline src/lib/application.ts:114]: I think `validateProject` might be better - it immediately implies that the callback will be passed a project, thoughts?

Comment (RebeccaStevens) [reply src/lib/application.ts:114]: So `ApplicationEvents.VALIDATE_PROJECT`?

Comment (Gerrit0) [reply src/lib/application.ts]: Yep, and `validateProject` for the string, otherwise lgtm

Comment (RebeccaStevens) [reply src/lib/application.ts]: Thanks all set up :)

**Comment by Gerrit0** [comment]
Thanks! Should release today.

**Comment by Gerrit0** [comment]
> When working on my plugin that relies on this PR, reflection.comment.getTags("@param") always return an empty array. I'm not having this issue with any other tags. Am I missing something for this tag or is this a bug?

Just realized I forgot to answer this -- `@param` tags are moved to the `ParameterReflection` instances, so you generally shouldn't have any `@param` tags, your plugin should probably check that `signatureReflection.parameters` all have comments.

**Pull request merged**